File: fight/fight.ml
type state = UNFINISHED | VICTORY_OF of Fighter.team | DRAW
type status = {
    fighters : Fighter.fighter list;
    mutable turn : int
}
let max_turns = 100

(* Returns the state of the fight. *)
let fightState (fighters : Fighter.fighter list) = match Fighter.getAlive (Fighter.getTeam Fighter.ALLY fighters) with
    | [] -> (match Fighter.getAlive (Fighter.getTeam Fighter.ENEMY fighters) with
        | [] -> DRAW
        | _ -> VICTORY_OF Fighter.ENEMY)
    | _ -> (match Fighter.getAlive (Fighter.getTeam Fighter.ENEMY fighters) with
        | [] -> VICTORY_OF Fighter.ALLY
        | _ -> UNFINISHED)

let rec doActions (order : Fighter.fighter list) (status : status) = match order with
    | x :: q -> begin
            x#debug;
            print_string "\t• Action\n";
            (* TODO: Determine active rule *)
            (* TODO: Apply the corresponding action *)
            print_string "\n";
            doActions q status
        end
    | [] -> ()

(* Turn logic, returns the state of the fight at the end of the turn. *)
let turn status = print_string ("Turn " ^ (string_of_int status.turn) ^ ":\n") ; match status.turn with
    | x when x = max_turns -> DRAW
    | _ -> begin
            doActions (Fighter.sort_invert status.fighters Rules.SPD) status;
            fightState status.fighters
        end

(* Returns 1 if f1 wins, -1 if f2 wins, or 0 in case of draw. *)
let fight f1 f2 =
    (* Prepare teams *)
    List.iter (fun f -> (f#prepare Fighter.ALLY)) f1;
    List.iter (fun f -> (f#prepare Fighter.ENEMY)) f2;
    let status = {fighters = (f2 @ f1); turn = 1} in
    (* Fight! *)
    let rec fight_turn status = match turn status with
        | UNFINISHED -> begin
                status.turn <- status.turn + 1;
                fight_turn status
            end
        | VICTORY_OF Fighter.ALLY -> 1
        | VICTORY_OF Fighter.ENEMY -> -1
        | DRAW -> 0
    in fight_turn status

File: battle-logic/src/logic/fight.rs
use std::cell::{Ref, RefCell, RefMut};
use std::ops::{Deref, DerefMut};

use crate::logic::prelude::*;

#[derive(Debug, PartialEq)]
pub enum State {
    AlliesVictory,
    EnemiesVictory,
    Draw,
}

#[derive(Debug, PartialEq, Copy, Clone)]
pub enum FighterID {
    Ally(usize),
    Enemy(usize),
    None,
}

impl FighterID {
    pub fn is_ally(&self) -> bool {
        match self {
            FighterID::Ally(_) => true,
            _ => false,
        }
    }
}

pub const MAX_TURNS: u8 = 50;

pub struct Fight {
    pub turn: u8,
    pub fighters: Vec<(FighterID, RefCell<Fighter>)>,
}

impl Fight {
    pub fn start(team1: Vec<Fighter>, team2: Vec<Fighter>) -> State {
        let mut fight = Fight::build_fight(team1, team2);

        loop {
            match fight.turn() {
                Some(result) => return result,
                None => (),
            }
        }
    }

    pub fn build_fight(team1: Vec<Fighter>, team2: Vec<Fighter>) -> Fight {
        let mut fighters = Vec::new();
        team1
            .into_iter()
            .enumerate()
            .for_each(|(i, f)| fighters.push((FighterID::Ally(i), RefCell::new(f))));
        team2
            .into_iter()
            .enumerate()
            .for_each(|(i, f)| fighters.push((FighterID::Enemy(i), RefCell::new(f))));

        Fight { turn: 0, fighters }
    }

    pub fn turn(&mut self) -> Option<State> {
        self.turn += 1;
        if self.turn > MAX_TURNS {
            return Some(State::Draw);
        }

        println!("Turn {}", self.turn);

        // Order fighters by speed
        self.fighters.sort_by_key(|(_, fighter)| fighter.borrow().deref().get_stat(&Stat::Speed));
        self.fighters.reverse();

        // Get turns order
        let mut turn_order = Vec::new();
        self.fighters
            .iter()
            .for_each(|(id, _)| turn_order.push(id.clone()));

        let mut state: Option<State> = None;

        for id in turn_order {
            let (action, target) = {
                let mut active = self.get_fighter_mut(id);
                if !active.is_alive() { continue; };

                // Start of turn logic
                active.turn();

                // Resolve rule, action, target for the turn
                let rule = active.get_rule(&self);
                let action = rule.action.clone();
                let target = action.get_target(&id, &self);
                (action, target)
            };

            let consequences = {
                action.execute(self.get_fighter(id).deref(), self.get_fighter(target).deref())
            };

            for (on_self, consequence) in consequences {
                match on_self {
                    WeaponTarget::Me => consequence.apply_on(self.get_fighter_mut(id).deref_mut()),
                    WeaponTarget::Other => consequence.apply_on(self.get_fighter_mut(target).deref_mut()),
                };
            }

            state = self.check_state();
            if state != None {
                break;
            }
        }

        return state;
    }

    fn get_fighter(&self, id: FighterID) -> Ref<Fighter> {
        self.fighters.iter().find(|(f_id, _)| *f_id == id).unwrap().1.borrow()
    }

    fn get_fighter_mut(&self, id: FighterID) -> RefMut<Fighter> {
        self.fighters.iter().find(|(f_id, _)| *f_id == id).unwrap().1.borrow_mut()
    }

    pub fn check_state(&self) -> Option<State> {
        if self.fighters
            .iter()
            .filter(|(id, _)| !id.is_ally())
            .all(|(_, f)| !f.borrow().is_alive())
        {
            return Some(State::AlliesVictory);
        } else if self.fighters
            .iter()
            .filter(|(id, _)| id.is_ally())
            .all(|(_, f)| !f.borrow().is_alive())
        {
            return Some(State::EnemiesVictory);
        }
        return None;
    }
}
